modules/host/network.go

```diff
@@ -20,57 +21,116 @@ func (h *Host) initNetworking(address string) error {
 	}
 	h.netAddress = modules.NetAddress(h.listener.Addr().String())
 
-	// Networking subroutines.
-	go h.forwardPort(h.netAddress.Port())
-	go h.learnHostname()
-	go h.listen()
-	return nil
-}
+	// Non-blocking, perform port forwarding and hostname discovery.
+	go func() {
+		h.resourceLock.RLock()
+		defer h.resourceLock.RUnlock()
 
-// listen listens for incoming RPCs and spawns an appropriate handler for each.
-func (h *Host) listen() {
-	for {
-		conn, err := h.listener.Accept()
-		if err != nil {
+		// If the host has closed, return immediately.
+		if h.closed {
 			return
 		}
-		go h.handleConn(conn)
-	}
+
+		h.mu.RLock()
+		port := h.netAddress.Port()
+		h.mu.RUnlock()
+		h.forwardPort(port)
+		h.learnHostname()
+	}()
+
+	// Launch the listener.
+	h.resourceLock.RLock()
+	go h.threadedListen()
+	return nil
 }
 
-// handleConn handles an incoming connection to the host, typically an RPC.
-func (h *Host) handleConn(conn net.Conn) {
-	defer conn.Close()
+// threadedHandleConn handles an incoming connection to the host, typically an
+// RPC. threadedHandleConn is responsible for releasing a readlock on
+// host.resourceLock when all communications have completed.
+func (h *Host) threadedHandleConn(conn net.Conn) {
+	// All threaded functions are called holding the close lock, and are
+	// expected to keep it until they no longer need access to the host's
+	// resources.
```

Comment:
wat??? why? This seems like a dangerous convention. Locking->Unlocking across function boundaries = bad plan